### PR TITLE
Move "assessment", "named_level" and "bonus" into the properties hash

### PR DESCRIPTION
**Pull request by joshlory**
Followup to https://github.com/code-dot-org/code-dot-org/pull/14728#issuecomment-298068493 and https://github.com/code-dot-org/code-dot-org/pull/15179.  Migrate `ScriptLevel` boolean flags to a serialized attribute enum.

### CODE REVIEW

File: dashboard/db/migrate/20170518002933_remove_columns_from_script_levels.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      when ScriptLevel::NAMED_LEVEL
+        sl.update!(named_level: true)
+      when ScriptLevel::ASSESSMENT
+        sl.update!(assessment: true)
+      when ScriptLevel::BONUS
+        sl.update!(bonus: true)
+      end
+    end
+  end
+end
```
Comment (Bjvanminnen) [inline dashboard/db/migrate/20170518002933_remove_columns_from_script_levels.rb:37]: Channeling my inner Asher, are we able to separate migration from code changes (I'm not actually sure how we'd accomplish that here).

Comment (joshlory) [reply dashboard/db/migrate/20170518002933_remove_columns_from_script_levels.rb:37]: We could add the type enum and start updating both places, but it leaves us in a weird state until we drop the old columns and update the tests.  I've verified the migration is reversible and preserves the column data.

Comment (balderdash) [reply dashboard/db/migrate/20170518002933_remove_columns_from_script_levels.rb:37]: The script cache should mostly insulate us from errors in the time between migrating the DB and upgrading the frontends, so this won't be as bad as coupling the migration and code changes would normally be. Maybe just give DotD a heads up when this PR is going to go to prod?

**Comment by joshlory** [comment]
@wjordan or @ashercodeorg, any concerns with this migration of data from boolean columns to a property enum?

I included the migration in this PR to avoid writing the data in two different places which could get out sync.  I've verified the migration is reversible and preserves the column data.

File: dashboard/app/models/script_level.rb [hunk 2]
```diff
@@ -41,13 +38,30 @@ class ScriptLevel < ActiveRecord::Base
...
+    type == NAMED_LEVEL
+  end
+
+  def assessment
+    type == ASSESSMENT
+  end
+
+  def bonus
+    type == BONUS
+  end
```
Comment (ashercodeorg) [inline dashboard/app/models/script_level.rb:63]: Should these be named `named_level?`, `assessment?`, and `bonus?`?

Comment (joshlory) [reply dashboard/app/models/script_level.rb:63]: These are replicating the existing boolean column accessors that I'm removing.  In the future my plan is to update any code that does e.g. `script_level.bonus` with `script_level.type == bonus`.

File: dashboard/app/models/script_level.rb [hunk 3]
```diff
@@ -41,13 +38,30 @@ class ScriptLevel < ActiveRecord::Base
     progression
     target
     challenge
+    type
   )
 
+  NAMED_LEVEL = 'named_level'
+  ASSESSMENT = 'assessment'
+  BONUS = 'bonus'
```
Comment (ashercodeorg) [inline dashboard/app/models/script_level.rb:46]: nit: Use `freeze` on CONSTANTS.

**Comment by joshlory** [comment]
Potentially yes, however as Ram noted this should be covered by caching of Scripts and Script Levels on each FE.

**Comment by ashercodeorg** [comment]
That I wouldn't assume, for example I have an [outstanding PR](https://github.com/code-dot-org/code-dot-org/pull/15297) fixing a place where we aren't caching.

**Comment by joshlory** [comment]
I'm planning to update this PR to do (3), but that work will most likely happen at the end of this sprint / beginning of next sprint.  Thanks all for the detailed review!

**Pull request closed without merging**